analyses/1.Transfer_Sepsis_to_TB.ipynb

```diff
@@ -0,0 +1,425 @@
+   "execution_count": 11,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Load tuberculosis data\n",
+    "tb_dataset = dataset.ExpressionDataset(test_dirs, sample_to_label, label_to_encoding, gene_file)\n",
+    "tb_loader = torch.utils.data.DataLoader(tb_dataset, batch_size=16, num_workers=4, pin_memory=True)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Evaluate erm and irm models on tb set"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 12,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "erm_net = torch.load('../logs/sepsis_erm.pkl')\n",
+    "irm_net = torch.load('../logs/sepsis_irm.pkl')"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 25,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "erm_loss = 0\n",
+    "erm_correct = 0\n",
+    "irm_loss = 0\n",
+    "irm_correct = 0\n",
+    "\n",
+    "erm_tb_preds= 0\n",
+    "irm_tb_preds = 0\n",
+    "\n",
+    "\n",
+    "for tb_batch in tb_loader:\n",
+    "    expression, labels, ids = tb_batch                                            \n",
+    "    tune_expression = expression.to(device)                                         \n",
+    "    tune_labels = labels.to(device).double()                                        \n",
+    "\n",
+    "    loss_function = torch.nn.BCEWithLogitsLoss()                                          \n",
+    "\n",
+    "    erm_preds = erm_net(tune_expression)\n",
+    "    loss = loss_function(erm_preds, tune_labels)\n",
+    "    erm_loss += float(loss)\n",
+    "    erm_correct += util.count_correct(erm_preds, tune_labels)\n",
+    "    \n",
+    "    irm_preds = irm_net(tune_expression)      \n",
+    "    loss = loss_function(irm_preds, tune_labels)\n",
+    "    irm_loss += float(loss)\n",
+    "    irm_correct += util.count_correct(irm_preds, tune_labels)\n",
+    "    \n",
+    "    # Count how many times the model predicted something other than zero\n",
+    "    erm_tb_preds += np.sum(erm_preds.cpu().detach().numpy() >= 0)\n",
+    "    irm_tb_preds += np.sum(irm_preds.cpu().detach().numpy() >= 0)\n",
+    "    \n",
+    "avg_erm_loss = erm_loss / len(tb_dataset)\n",
+    "erm_acc = erm_correct / len(tb_dataset)\n",
+    "avg_irm_loss = irm_loss / len(tb_dataset)\n",
+    "irm_acc = irm_correct / len(tb_dataset)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "808\n",
+      "520\n",
+      "907\n"
+     ]
+    }
+   ],
+   "source": [
+    "print(erm_tb_preds)\n",
+    "print(irm_tb_preds)\n",
+    "print(len(tb_dataset))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 18,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "ERM trained network tb accuracy: 0.7375964718853363\n",
```

Comment:
Interesting that IRM seems to have a better (lower) loss value on the TB data, but the accuracy is quite a bit lower.

Reply:
This was surprising to me too, especially because ERM optimizes the loss function we're using as the metric here while IRM jointly optimizes the BCE loss and the gradient penalty. 

The ERM model is predicting mainly the dominant class, while the IRM class predictions are more balanced. This lowers the accuracy in IRM, but the decreased loss might indicate that IRM's representation is more biologically interesting